### PR TITLE
Dotnet ecosystem metric collection

### PR DESCRIPTION
**Pull request by sachin-sandhu**
### What are you trying to accomplish?

Dotnet ecosystem metric collection

Collects following:

Nuget package manager version information 
Dotnet languages C#, F#, VB metric along with targeted framework version
Ecosystem metric 

### Checklist

- [x] I have run the complete test suite to ensure all tests and linters pass.
- [x] I have thoroughly tested my code changes to ensure they work as expected, including adding additional tests for new functionality.
- [x] I have written clear and descriptive commit messages.
- [x] I have provided a detailed description of the changes in the pull request, including the problem it addresses, how it fixes the problem, and any relevant details about the implementation.
- [x] I have ensured that the code is well-documented and easy to understand.

### CODE REVIEW

**Review by randhircs — APPROVED** [review]
To me looks good.

**Review by amazimbe — APPROVED** [review]

File: nuget/lib/dependabot/nuget/file_parser.rb [hunk 1]
```diff
@@ -22,8 +24,36 @@ def parse
...
+            package_manager: package_manager,
+            language: language
+          ),
+          T.nilable(Ecosystem)
+        )
+      end
+
       private
 
+      sig { returns(T.nilable(T::Array[T.nilable(String)])) }
```
Comment (JamieMagee) [inline nuget/lib/dependabot/nuget/file_parser.rb:41]: If you add a `.compact` to the end of your `.map` below, you can simplify this to a `T.nilable(T::Array[String])`

File: nuget/lib/dependabot/nuget/file_parser.rb [hunk 2]
```diff
@@ -22,8 +24,36 @@ def parse
...
+        @content_json ||= T.let(begin
+          directory = source&.directory || "/"
+          discovery_json_reader = DiscoveryJsonReader.run_discovery_in_directory(
+            repo_contents_path: T.must(repo_contents_path),
+            directory: directory,
+            credentials: credentials
+          )
+
+          discovery_json_reader.workspace_discovery&.projects&.map do |framework|
+            T.let(framework.instance_variable_get(:@target_frameworks), T::Array[String]).first
```
Comment (JamieMagee) [inline nuget/lib/dependabot/nuget/file_parser.rb:52]: .NET projects can target multiple frameworks, so you're potentially omitting anything but the first listed one here and your telemetry may not be giving an accurate picture.

File: nuget/lib/dependabot/nuget/file_parser.rb [hunk 3]
```diff
@@ -53,6 +83,65 @@ def check_required_files
...
+      end
+
+      sig { returns(Ecosystem::VersionManager) }
+      def package_manager
+        NugetPackageManager.new(T.must(nuget_version))
+      end
+
+      sig { returns(T.nilable(String)) }
+      def nuget_version
+        SharedHelpers.run_shell_command("dotnet nuget --version").split("Command Line").last&.strip
```
Comment (JamieMagee) [inline nuget/lib/dependabot/nuget/file_parser.rb:141]: I _think_ this is always going to return the same version as we install in the container. So don't be surprised if this always returns the same version.

File: nuget/lib/dependabot/nuget/file_parser.rb (hunk 1 above)
Comment (sachin-sandhu) [reply nuget/lib/dependabot/nuget/file_parser.rb]: @JamieMagee fixed it, thanks for the suggestion.

File: nuget/lib/dependabot/nuget/file_parser.rb (hunk 2 above)
Comment (sachin-sandhu) [reply nuget/lib/dependabot/nuget/file_parser.rb]: @JamieMagee , understood , i have merged the versions as a single string now.

File: nuget/lib/dependabot/nuget/file_parser.rb [hunk 4]
```diff
@@ -53,6 +83,65 @@ def check_required_files
           "No project file."
         )
       end
+
+      sig { returns(T.nilable(Ecosystem::VersionManager)) }
+      def language
+        # Historically new version of language is released with incremental update of
+        # .Net version, so we tie the language with framework version for metric collection
+
+        nomenclature = "#{language_type} #{framework_version&.first}".strip.tr(" ", "-")
```
[the remark this answers is not in the document]
Comment (sachin-sandhu) [reply nuget/lib/dependabot/nuget/file_parser.rb]: @JamieMagee , fixed it as per suggestion.

File: nuget/lib/dependabot/nuget/file_parser.rb (hunk 3 above)
Comment (sachin-sandhu) [reply nuget/lib/dependabot/nuget/file_parser.rb:141]: i was thinking the same.

**Review by JamieMagee — APPROVED** [review]

**Pull request merged**